File: backend/app/schemas/teacher.py
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field

from backend.app.schemas.curriculum import SubjectCode


class InteractionMode(str, Enum):
    STANDARD = "standard"
    FALLBACK_EXPLANATION = "fallback_explanation"
    ASSESSMENT_HINT = "assessment_hint"


class TeacherPersonaConfig(BaseModel):
    name: str = Field(default="Suman AI")
    language: str = Field(default="en", description="Target language: 'en', 'hi', or 'mr'")
    tone: str = Field(default="warm, supportive, patient")
    grade_level: int = Field(default=1)


class RichVisualCardTitles(BaseModel):
    english: str = Field(..., description="Entity name in English")
    hindi: str = Field(..., description="Entity name in Hindi")
    marathi: str = Field(..., description="Entity name in Marathi")


class RichVisualCardPayload(BaseModel):
    entity_name: str = Field(..., description="Primary entity name")
    image_url: str = Field(..., description="URL to high quality image/illustration")
    titles: RichVisualCardTitles = Field(..., description="3-language name mappings")
    pronunciation_audio: Optional[str] = Field(
        None, description="URL or key for pronunciation audio"
    )
    simple_explanation: str = Field(..., description="Age-appropriate Class 1 explanation")
    checking_question: str = Field(..., description="Suman AI interactive checking question")
    class_level: int = Field(default=1, description="Target Class level")


class TeacherChatRequest(BaseModel):
    student_id: str = Field(..., description="Unique student/user identifier")
    message: str = Field(..., description="Student query or response")
    language: str = Field(default="en", description="Preferred language ('en', 'hi', 'mr')")
    subject_code: Optional[SubjectCode] = Field(None, description="Active subject context")
    topic_id: Optional[str] = Field(None, description="Active topic context")
    subtopic_id: Optional[str] = Field(None, description="Active subtopic context")
    concept_id: Optional[str] = Field(None, description="Active concept context")
    is_confused: bool = Field(
        default=False,
        description="Explicit flag if student clicks 'I don't understand' button",
    )


class TeacherChatResponse(BaseModel):
    response_text: str = Field(..., description="Suman AI Teacher response message")
    language: str = Field(..., description="Language of response")
    interaction_mode: InteractionMode = Field(
        default=InteractionMode.STANDARD, description="Mode of response generated"
    )
    is_fallback_explanation: bool = Field(
        default=False, description="True if response was generated via fallback engine"
    )
    visual_cue_trigger: Optional[str] = Field(
        None, description="Extracted visual cue graphic hint if present"
    )
    rich_card: Optional[RichVisualCardPayload] = Field(
        None, description="Multi-Modal Rich Visual Card payload if entity match exists"
    )
    grounded_concept_id: Optional[str] = Field(
        None, description="Concept ID response was grounded on"
    )
    safety_filtered: bool = Field(
        default=False, description="True if response/query triggered child safety filter"
    )
    encouragement_phrase: Optional[str] = Field(None, description="Warm supportive phrase")
    simple_explanation: Optional[str] = Field(
        None, description="Kid-friendly simple conceptual explanation"
    )
    real_life_example: Optional[str] = Field(
        None, description="Relatable everyday real-life example"
    )
    checking_question: Optional[str] = Field(
        None, description="Interactive checking question to verify comprehension"
    )
    child_response_evaluation: Optional[str] = Field(
        None, description="Evaluation feedback on child's previous attempt"
    )
    supportive_correction: Optional[str] = Field(
        None, description="Warm supportive correction and guidance"
    )
    teaching_strategy: str = Field(
        default="standard_analogy",
        description="Strategy: 'story', 'simpler_analogy', 'visual', or 'standard_analogy'",
    )

File: mobile/lib/models/teacher.dart
import 'package:nehal_teacher_ai_mobile/widgets/rich_visual_card.dart';

enum InteractionMode {
  standard,
  fallbackExplanation,
  assessmentHint,
}

class TeacherChatMessage {
  final String id;
  final String sender; // 'student' or 'teacher'
  final String messageText;
  final DateTime timestamp;
  final bool isFallbackExplanation;
  final String? visualCueTrigger;
  final RichVisualCardData? richCard;
  final String? simpleExplanation;
  final String? realLifeExample;
  final String? checkingQuestion;
  final String? childResponseEvaluation;
  final String? supportiveCorrection;
  final String teachingStrategy;

  TeacherChatMessage({
    required this.id,
    required this.sender,
    required this.messageText,
    required this.timestamp,
    this.isFallbackExplanation = false,
    this.visualCueTrigger,
    this.richCard,
    this.simpleExplanation,
    this.realLifeExample,
    this.checkingQuestion,
    this.childResponseEvaluation,
    this.supportiveCorrection,
    this.teachingStrategy = 'standard_analogy',
  });

  factory TeacherChatMessage.fromJson(Map<String, dynamic> json) {
    return TeacherChatMessage(
      id: json['id'] as String? ?? DateTime.now().millisecondsSinceEpoch.toString(),
      sender: json['sender'] as String? ?? 'teacher',
      messageText: json['message_text'] as String,
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'] as String)
          : DateTime.now(),
      isFallbackExplanation: json['is_fallback_explanation'] as bool? ?? false,
      visualCueTrigger: json['visual_cue_trigger'] as String?,
      richCard: json['rich_card'] != null
          ? RichVisualCardData.fromJson(json['rich_card'] as Map<String, dynamic>)
          : null,
      simpleExplanation: json['simple_explanation'] as String?,
      realLifeExample: json['real_life_example'] as String?,
      checkingQuestion: json['checking_question'] as String?,
      childResponseEvaluation: json['child_response_evaluation'] as String?,
      supportiveCorrection: json['supportive_correction'] as String?,
      teachingStrategy: json['teaching_strategy'] as String? ?? 'standard_analogy',
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'sender': sender,
        'message_text': messageText,
        'timestamp': timestamp.toIso8601String(),
        'is_fallback_explanation': isFallbackExplanation,
        'visual_cue_trigger': visualCueTrigger,
        'simple_explanation': simpleExplanation,
        'real_life_example': realLifeExample,
        'checking_question': checkingQuestion,
        'child_response_evaluation': childResponseEvaluation,
        'supportive_correction': supportiveCorrection,
        'teaching_strategy': teachingStrategy,
      };
}

class TeacherChatRequest {
  final String studentId;
  final String message;
  final String language;
  final String? subjectCode;
  final String? topicId;
  final String? subtopicId;
  final String? conceptId;
  final bool isConfused;

  TeacherChatRequest({
    required this.studentId,
    required this.message,
    this.language = 'en',
    this.subjectCode,
    this.topicId,
    this.subtopicId,
    this.conceptId,
    this.isConfused = false,
  });

  Map<String, dynamic> toJson() => {
        'student_id': studentId,
        'message': message,
        'language': language,
        'subject_code': subjectCode,
        'topic_id': topicId,
        'subtopic_id': subtopicId,
        'concept_id': conceptId,
        'is_confused': isConfused,
      };
}

class TeacherChatResponse {
  final String responseText;
  final String language;
  final String interactionMode;
  final bool isFallbackExplanation;
  final String? visualCueTrigger;
  final RichVisualCardData? richCard;
  final String? groundedConceptId;
  final bool safetyFiltered;
  final String? encouragementPhrase;
  final String? simpleExplanation;
  final String? realLifeExample;
  final String? checkingQuestion;
  final String? childResponseEvaluation;
  final String? supportiveCorrection;
  final String teachingStrategy;

  TeacherChatResponse({
    required this.responseText,
    required this.language,
    required this.interactionMode,
    this.isFallbackExplanation = false,
    this.visualCueTrigger,
    this.richCard,
    this.groundedConceptId,
    this.safetyFiltered = false,
    this.encouragementPhrase,
    this.simpleExplanation,
    this.realLifeExample,
    this.checkingQuestion,
    this.childResponseEvaluation,
    this.supportiveCorrection,
    this.teachingStrategy = 'standard_analogy',
  });

  factory TeacherChatResponse.fromJson(Map<String, dynamic> json) {
    return TeacherChatResponse(
      responseText: json['response_text'] as String,
      language: json['language'] as String,
      interactionMode: json['interaction_mode'] as String? ?? 'standard',
      isFallbackExplanation: json['is_fallback_explanation'] as bool? ?? false,
      visualCueTrigger: json['visual_cue_trigger'] as String?,
      richCard: json['rich_card'] != null
          ? RichVisualCardData.fromJson(json['rich_card'] as Map<String, dynamic>)
          : null,
      groundedConceptId: json['grounded_concept_id'] as String?,
      safetyFiltered: json['safety_filtered'] as bool? ?? false,
      encouragementPhrase: json['encouragement_phrase'] as String?,
      simpleExplanation: json['simple_explanation'] as String?,
      realLifeExample: json['real_life_example'] as String?,
      checkingQuestion: json['checking_question'] as String?,
      childResponseEvaluation: json['child_response_evaluation'] as String?,
      supportiveCorrection: json['supportive_correction'] as String?,
      teachingStrategy: json['teaching_strategy'] as String? ?? 'standard_analogy',
    );
  }

  Map<String, dynamic> toJson() => {
        'response_text': responseText,
        'language': language,
        'interaction_mode': interactionMode,
        'is_fallback_explanation': isFallbackExplanation,
        'visual_cue_trigger': visualCueTrigger,
        'grounded_concept_id': groundedConceptId,
        'safety_filtered': safetyFiltered,
        'encouragement_phrase': encouragementPhrase,
        'simple_explanation': simpleExplanation,
        'real_life_example': realLifeExample,
        'checking_question': checkingQuestion,
        'child_response_evaluation': childResponseEvaluation,
        'supportive_correction': supportiveCorrection,
        'teaching_strategy': teachingStrategy,
      };
}

class TeacherPersonaConfig {
  final String name;
  final String language;
  final String tone;
  final int gradeLevel;

  TeacherPersonaConfig({
    required this.name,
    required this.language,
    required this.tone,
    required this.gradeLevel,
  });

  factory TeacherPersonaConfig.fromJson(Map<String, dynamic> json) {
    return TeacherPersonaConfig(
      name: json['name'] as String? ?? 'Suman AI',
      language: json['language'] as String? ?? 'en',
      tone: json['tone'] as String? ?? 'warm, supportive, patient',
      gradeLevel: json['grade_level'] as int? ?? 1,
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'language': language,
        'tone': tone,
        'grade_level': gradeLevel,
      };
}
